Add scaffolding for aoti_torch_call_dispatcher BC with native ops

mikaylagawarecki opened the pull request:
Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #165285
* #165284
* #165283
* #164332
* __->__ #163683
* #166373
* #164356

mikaylagawarecki committed:
Add scaffolding for aoti_torch_call_dispatcher BC/FC with native ops

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC/FC with native ops"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC/FC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_abi_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC/FC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_abi_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC/FC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_abi_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC/FC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_abi_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_abi_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_abi_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `build_abi_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes




[ghstack-poisoned]

janeyx99 commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -537,6 +537,17 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_call_dispatcher(
+// Has the same semantic as aoti_torch_call_dispatcher, but takes an
+// additional argument for the extension build version. This is
+// needed for backward compatibility when calling native functions via
+// the dispatcher. The caller should pass in its build version (not target
+// version).
```
```suggestion
// the dispatcher. The caller should pass in the libtorch version the extension is
// building with (NOT target version).
```
making sure I understand

janeyx99 commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -537,6 +537,17 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_call_dispatcher(
+    const char* opName,
+    const char* overloadName,
+    StableIValue* stack,
+    uint64_t extension_build_version);
```
My conclusion from reading this comment is:
- TORCH_FEATURE_VERSION is used in our headers to bar code that calls into shim APIs that do not exist for a certain version.
- TORCH_ABI_VERSION is the "real" status of APIs, so we need to adapt based on reality (the struct or API at the time of build) vs an older version of reality.

In this new framework, we are relying on TORCH_FEATURE_VERSION to be the gatekeeper. Once we're "through the gate", we assume that the APIs we are using are what we have, so we use TORCH_ABI_VERSION to communicate that information.

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR




[ghstack-poisoned]

janeyx99 commented on test/cpp_extensions/schema_adapter_test_extension/schema_adapter_test/csrc/test_extension.cpp:
```diff
@@ -0,0 +1,151 @@
+
+// Pybind11 module - test adapters are automatically registered via static
+// initialization
+PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
```
do we need this?

janeyx99 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1672,6 +1675,234 @@ AOTITorchError aoti_torch_call_dispatcher(
+
+    // there should then be num_returns IValues on the stack, which
+    // we will convert to StableIValue and repopulate user input stack
+    for (const auto idx : c10::irange(num_returns)) {
```
I'd anticipate there is a future possibility we may need adapters for returns too. From what I understand, the approach should be the same as with the inputs and thus can be added later without changing any APIs?

janeyx99 reviewed (approved):
Thanks! LGTM. According to our in person discussion I also agree we should move out the testing code.

mikaylagawarecki commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1672,6 +1675,234 @@ AOTITorchError aoti_torch_call_dispatcher(
+// change**. If there is a need to define similar infrastructure for the returns
+// we can update this.
+
+namespace {
```
Yep!

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add scaffolding for aoti_torch_call_dispatcher BC with native ops"

Part 1 of plan in https://docs.google.com/document/d/1MaX51H5aEQE5XnOlnZIpf9oCYwzGrTWkgBACxNzsmWE/edit?usp=sharing

- Upgrade `aoti_torch_call_dispatcher` to v2 with an `extension_build_version`
- Allow registration of StableIValue stack  --> IValue stack adapters for schema changes

#### Note: This PR does not include a linter that tells the user to add the upgrader if the schema changes, which is an important piece that will be added in a separate PR




[ghstack-poisoned]
